File: Chapter 07/JS/section_7_4_event_code/core/eventPublisher.js
/**
 * THE ABSTRACTION: This keeps your domain logic completely decoupled.
 * The Controller will depend on this abstraction so it has no idea whether 
 * it is publishing to an RxJS Subject, a RabbitMQ cluster, or AWS EventBridge.
 */
export class EventPublisher {
    /**
     * @param event - The event to publish into the stream.
     */
    publish(event) {
        throw new Error("Method 'publish()' must be implemented by the infrastructure layer.");
    }
}

File: Chapter 07/Python/section_7_4_event_code/core/event_publisher.py
from abc import ABC, abstractmethod
from ..shared.event import Event

class EventPublisher(ABC):
    """
    THE ABSTRACTION: This keeps your domain logic completely decoupled.
    
    The Controller will depend on this abstraction so it has no idea whether 
    it is publishing to an asyncio Queue, a RabbitMQ cluster, or AWS EventBridge.
    It only knows that when it calls publish_async, the system will handle it.
    """
    
    @abstractmethod
    async def publish_async(self, event: Event) -> None:
        pass